### PR TITLE
Dictionaries, CallStack, CSPR Deposits

### PR DESCRIPTION
**Pull request by zie1ony**

### CODE REVIEW

File: account-info/src/lib.rs [hunk 1]
```diff
@@ -193,20 +253,18 @@ fn set_url_for_account() {
...
 /// Administrator function to remove stored data from the contract.
 #[no_mangle]
 fn delete_url_for_account() {
     Admins::new().assert_caller_is_admin();
 
-    let public_key = runtime::get_named_arg::<PublicKey>("public_key");
-    runtime::remove_key(&pubkey_to_string(&public_key));
+    let account = runtime::get_named_arg("account");
+    Urls::new().delete(&account);
+    Deposits::new().withdraw(&account);
```
Comment (ihor) [inline account-info/src/lib.rs:267]: @zie1ony, since the ```set_url_for_account``` doesn't deposit any amount, it will not be possible to delete URL set by admins because ```Deposits::new().withdraw(&account);``` will revert the changes with an error. Right?

File: account-info/src/lib.rs [hunk 2]
```diff
@@ -164,26 +204,46 @@ pub fn install_or_upgrade_contract(name: String) {
...
     runtime::ret(CLValue::from_t(url).unwrap_or_revert());
 }
 
 /// Function so the caller can remove their stored URL from the contract.
 #[no_mangle]
 fn delete_url() {
-    runtime::remove_key(&runtime::get_caller().to_string());
+    let caller = utils::get_caller();
+    Urls::new().delete(&caller);
+    Deposits::new().withdraw(&caller);
```
Comment (ihor) [inline account-info/src/lib.rs:246]: Similarly here, if the URL was set by an admin ```Deposits::new().withdraw(&caller);``` will revert the changes with an error.

File: account-info/src/deposits.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    pub fn set_deposit_amount(&self, amount: U512) {
+        utils::set_key(DEPOSIT_AMOUNT, amount);
+    }
+
+    fn get_deposit(&self, account: &AccountHash) -> U512 {
+        storage::dictionary_get(self.dict_uref, &account.to_string())
+            .unwrap_or_default()
+            .unwrap_or_default()
```
Comment (ihor) [inline account-info/src/deposits.rs:65]: The code line above seems like unnecessary duplication

File: account-info/src/urls.rs [hunk 4]
```diff
@@ -1,3 +1,44 @@
...
+    }
+    
+    pub fn delete(&self, address: &AccountHash) {
+        storage::dictionary_put(self.dict_uref, &address.to_string(), "");
+    }
+
+    pub fn get(&self, address: &AccountHash) -> String {
+        let url: String = storage::dictionary_get(self.dict_uref, &address.to_string())
+            .unwrap_or_revert()
+            .unwrap_or_revert_with(ContractError::NotFound);
```
Comment (ihor) [inline account-info/src/urls.rs:34]: It looks like on the former out of the two lines above is needed (unless I'm missing something)

File: account-info/src/urls.rs [hunk 5]
```diff
@@ -1,3 +1,44 @@
...
+
+    pub fn set(&self, address: &AccountHash, url: &str) {
+        if !is_valid_url(url) {
+            runtime::revert(ContractError::BadUrlFormat);
+        }
+        storage::dictionary_put(self.dict_uref, &address.to_string(), url);
+    }
+    
+    pub fn delete(&self, address: &AccountHash) {
+        storage::dictionary_put(self.dict_uref, &address.to_string(), "");
```
Comment (ihor) [inline account-info/src/urls.rs:28]: Shouldn't we delete the key here instead of setting the value to an empty string?

File: account-info/src/call_set_url.rs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#![no_main]
+
```
Comment (ihor) [inline account-info/src/call_set_url.rs:2]: Is this module needed?

File: account-info/src/lib.rs [hunk 7]
```diff
@@ -164,26 +204,46 @@ pub fn install_or_upgrade_contract(name: String) {
...
+    let caller = utils::get_caller();
     let url: String = runtime::get_named_arg("url");
-
-    if !check_url(&url) {
-        revert(ContractError::BadUrlFormat)
-    }
-    set_key(&runtime::get_caller().to_string(), url);
+    let purse: Option<URef> = runtime::get_named_arg("purse");
+    Deposits::new().deposit_if_needed(&caller, purse);
+    Urls::new().set(&caller, &url);
```
Comment (ihor) [inline account-info/src/lib.rs:211]: It's not clear if anyone outside of the contract supposed to call this entry point. Can it be merged with ```call_set_url```? 

Additionally, the ```purse``` argument is not defined in ```get_entry_points```. That confused me.

File: account-info/src/admins.rs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    pub fn add(&self, address: &AccountHash) {
+        if self.is_admin(address) {
+            runtime::revert(ContractError::AdminExists);
+        } else {
+            storage::dictionary_put(self.dict_uref, &address.to_string(), ADMIN_ACTIVE);
+            
+            // Increment the admin count.
+            let admins_count: u32 = utils::get_key(ADMINS_COUNT);
+            utils::set_key(ADMINS_COUNT, admins_count + 1);
```
Comment (ihor) [inline account-info/src/admins.rs:43]: Could you please confirm that the deploys are executed one by one. I remember that Ed said it is, but want to double check that there was no misunderstanding on my side. Thanks!

File: account-info/src/lib.rs [hunk 9]
```diff
@@ -1,25 +1,22 @@
 extern crate alloc;
```
Comment (ihor) [inline account-info/src/lib.rs:2]: Not related to this PR specifically, but since multiple modules were added, should this module be called ```lib.rs```? What is the convention?  ```contract.rs``` seems to be a much more descriptive name. What do you think?

File: account-info/src/call_set_url.rs (hunk 6 above)
Comment (deuszex) [reply account-info/src/call_set_url.rs:2]: Do you mean the file or the `#![no_main]` ?

File: account-info/src/urls.rs (hunk 5 above)
Comment (deuszex) [reply account-info/src/urls.rs:28]: Right now there is no function that deletes a key from a dictionary.

File: account-info/src/deposits.rs (hunk 3 above)
Comment (deuszex) [reply account-info/src/deposits.rs:65]: @ihor Could you be a bit more specific?
If you mean the double unwrap, it is because dictionary_get returns with Result<Option<V>, E>, so you have to unwrap once the Result to check whether the dictionary throws an error, then once again to see if the queried key held a value or not.

File: account-info/src/lib.rs (hunk 1 above)
Comment (deuszex) [reply account-info/src/lib.rs:267]: That is also how I read it.

File: account-info/src/call_set_url.rs (hunk 6 above)
Comment (ihor) [reply account-info/src/call_set_url.rs:2]: @deuszex the file itself seems to redundant. Or maybe I'm missing something.

File: account-info/src/deposits.rs (hunk 3 above)
Comment (ihor) [reply account-info/src/deposits.rs:65]: Thanks @deuszex! That wasn't clear to me by reading the code.

File: account-info/src/urls.rs (hunk 5 above)
Comment (ihor) [reply account-info/src/urls.rs:28]: Got it!

File: account-info/src/urls.rs (hunk 4 above)
Comment (ihor) [reply account-info/src/urls.rs:34]: This one is no longer relevant after @deuszex educated me that dictionary_get returns with Result<Option, E> :)
